simpleui-server/src/SuiZMQ.ts

```diff
@@ -0,0 +1,150 @@
+/**
+ * Created by Zack Beucler
+ * 
+ * potenitally add a send timeout ZMQ_SNDTIMEO
+ *                      http://api.zeromq.org/3-0:zmq-setsockopt
+ *      might be good to add because zmq will wait forever if not told otherwise
+ * 
+ * 
+ ****************************** NOTE ******************************
+ *
+ *  An http request can only request data from a single zmq socket
+ *
+ ******************************************************************
+ */
+
+
+var _zmq = require('zeromq');
+import {ServerUtil} from './server-util';
+import {Logger, LogLevel} from './server-logger';
+import { SuiData } from './sui_data';
+import { Queue } from './queue';
+
+
+export class ZMQ_Socker_Wrapper {
+    hostname: string;
+    timeout: number;
+    port: number;
+    socket: any;
+    http_queue: Queue;
+    
+
+    constructor(hostname: string='svcmachineapps', port: number, timeout: number=500) {
+        this.hostname = hostname;
+        this.timeout = timeout;
+        this.port = port;
+        this.http_queue = new Queue();
+
+        try {
+            this.socket = _zmq.socket('req');
+            this.socket.connect_timeout = timeout;
+            this.connect(this.port);
+
+            this.socket.on('message', (msg) => {
+                const raw_zmq_data = msg.toString();
+                const zmq_data = SuiData.addXmlStatus(raw_zmq_data);
+                Logger.log(LogLevel.DEBUG, `Recieved ZMQ message at port ${this.port}: ${zmq_data.substring(0, 105)}`);
+                // get response + request from queue
+                let [res, req] = this.http_queue.dequeue();
+                // send response
+                SuiData.sendResponse(req, res, zmq_data);
+            });
+
+            this.socket.on('error', (zmqErr) => {
+                const zmq_data = ServerUtil.getServerError('ZMQ_ERROR', '{{ERROR}}', zmqErr);
+                Logger.log(LogLevel.ERROR, `ZMQ socket at port ${this.port} got error: ${zmqErr}`);
+                // get response + request from queue
+                let [res, req] = this.http_queue.dequeue();
+                // send response
+                SuiData.sendResponse(req, res, zmq_data);
+            });
+
+        } catch (err) {
+            Logger.log(LogLevel.ERROR, `Could not create ZMQ socket at port ${this.port} got error: ${err}`)
```

Comment:
If the server can't construct the zmq socket then you should probably raise this to a higher level in the program (and, realistically, the program probably can't continue in this case and should just quit)

Reply:
i added a `process.exit(1)` after the error log line. Would it be better to throw an exception?

Reply:
In general it's better to throw an exception, then some higher level of the system can choose whether to catch it and terminate, catch it and ignore it, retry, etc. That said though, since you're building the whole stack here (as opposed to a library) you can get away with just calling terminate.